plugin/path_checkout_sets.go

```diff
@@ -114,41 +128,36 @@ func (b *backend) operationReserveCreate(ctx context.Context, req *logical.Reque
 		return logical.ErrorResponse(`"service_account_names" must be provided`), nil
 	}
 	// Ensure these service accounts aren't already managed by another check-out set.
-	var alreadyManagedErr error
+	var userErrs error
+	var resultingServiceAccountNames []string
 	for _, serviceAccountName := range serviceAccountNames {
-		if _, err := b.checkOutHandler.Status(ctx, req.Storage, serviceAccountName); err != nil {
-			if err == ErrNotFound {
-				// This is what we want to see.
+		if isUserErr, err := b.checkInNewServiceAccount(ctx, req.Storage, serviceAccountName); err != nil {
+			if isUserErr {
+				userErrs = multierror.Append(userErrs, err)
 				continue
 			}
-			// There is a more persistent error reaching storage.
-			return nil, err
-		}
-		// If we reach here, the error is nil. That means there's an existing CheckOut for this
-		// service account.
-		alreadyManagedErr = multierror.Append(alreadyManagedErr, fmt.Errorf("%s is already managed by another set, please remove it and try again", serviceAccountName))
-	}
-	if alreadyManagedErr != nil {
-		return logical.ErrorResponse(alreadyManagedErr.Error()), nil
-	}
-
-	// Now we need to check in all these service accounts so they'll be listed as managed by this
-	// plugin and available.
-	for _, serviceAccountName := range serviceAccountNames {
-		if err := b.checkOutHandler.CheckIn(ctx, req.Storage, serviceAccountName); err != nil {
 			return nil, err
 		}
+		resultingServiceAccountNames = append(resultingServiceAccountNames, serviceAccountName)
 	}
 
 	set := &librarySet{
-		ServiceAccountNames:       serviceAccountNames,
+		ServiceAccountNames:       resultingServiceAccountNames,
 		TTL:                       ttl,
 		MaxTTL:                    maxTTL,
 		DisableCheckInEnforcement: disableCheckInEnforcement,
 	}
+	if err := set.Validate(); err != nil {
+		return logical.ErrorResponse(err.Error()), nil
+	}
 	if err := storeSet(ctx, req.Storage, setName, set); err != nil {
 		return nil, err
 	}
+	if userErrs != nil {
+		// Let's return 400 here because we need to flag the user's attention
+		// that we didn't complete every requested action.
+		return logical.ErrorResponse(userErrs.Error()), nil
```

Comment:
It seems odd that we're returning a 400 error while having (possibly) successfully created a set. If we do want the request to succeed when there are user errors (not sure, per my question above), then returning these as warnings would be more appropriate. As a user I'd interpret 4/5xx and not succeeding at all.

Reply:
Yeah, if we proceed with the partial update approach, it's definitely a judgement call. I tried adding errors to Vault 200 response, but at least in the CLI, the errors get stripped out and it just looks like a vanilla 200. I could get around that with warnings.

Reply:
This is resolved thanks to no longer doing partial updates.